src/mpi/coll/bcast/bcast_intra_smp.c

```diff
@@ -49,87 +73,33 @@ int MPIR_Bcast_intra_smp(void *buffer, MPI_Aint count, MPI_Datatype datatype, in
     if (nbytes == 0)
         goto fn_exit;   /* nothing to do */
 
-    if ((nbytes < MPIR_CVAR_BCAST_SHORT_MSG_SIZE) || (comm_size < MPIR_CVAR_BCAST_MIN_PROCS)) {
-        /* send to intranode-rank 0 on the root's node */
-        if (local_root > 0) {   /* is not the node root (0) and is on our node (!-1) */
-            if (local_rank == local_root) {
-                mpi_errno = MPIC_Send(buffer, count, datatype, 0,
-                                      MPIR_BCAST_TAG, comm_ptr, node_group, errflag);
-                MPIR_ERR_CHECK(mpi_errno);
-            } else if (0 == comm_ptr->node_comm->rank) {
-                mpi_errno = MPIC_Recv(buffer, count, datatype,
-                                      MPIR_Get_intranode_rank(comm_ptr, root),
-                                      MPIR_BCAST_TAG, comm_ptr, node_group, status_p);
-                MPIR_ERR_CHECK(mpi_errno);
-#ifdef HAVE_ERROR_CHECKING
-                /* check that we received as much as we expected */
-                MPIR_Get_count_impl(status_p, MPI_BYTE, &recvd_size);
-                MPIR_ERR_CHKANDJUMP2(recvd_size != nbytes, mpi_errno, MPI_ERR_OTHER,
-                                     "**collective_size_mismatch",
-                                     "**collective_size_mismatch %d %d",
-                                     (int) recvd_size, (int) nbytes);
-#endif
-            }
-
-        }
-
+    if ((nbytes < MPIR_CVAR_BCAST_SHORT_MSG_SIZE) || (comm_size < MPIR_CVAR_BCAST_MIN_PROCS) ||
+        (nbytes < MPIR_CVAR_BCAST_LONG_MSG_SIZE && MPL_is_pof2(comm_size))) {
         /* local roots perform the internode broadcast */
-        if (local_rank == 0) {
+        if (local_rank == local_root) {
             mpi_errno = MPIR_Bcast(buffer, count, datatype, inter_root,
                                    comm_ptr, node_roots_group, errflag);
             MPIR_ERR_CHECK(mpi_errno);
         }
 
-        /* perform the intranode broadcast on all except for the root's node */
+        /* perform the intranode broadcast */
         if (local_size > 1) {
             mpi_errno = MPIR_Bcast(buffer, count, datatype, 0, comm_ptr, node_group, errflag);
             MPIR_ERR_CHECK(mpi_errno);
         }
```

Comment:
I definitely appreciate the simplicity of this approach vs the extra messaging. Can we capture any data to confirm this is not any slower vs the previous code with non-zero root?

Reply:
Here are some numbers measured using 4 nodes * 8 ppn
#### main branch
```
TEST bcast:
    :root=0
     msgsize      min      max      avg      sigma  (in microseconds)
           0    0.024    0.025    0.024      0.002
           1    0.279    6.151    4.429      0.317
           2    0.278    6.072    4.383      0.281
           4    0.282    6.171    4.411      0.301
           8    0.277    6.140    4.406      0.262
          16    0.279    6.178    4.428      0.295
          32    0.279    6.186    4.428      0.298
          64    0.278    6.200    4.458      0.270
         128    0.296    6.406    4.562      0.379
         256    2.989   14.814    8.276     46.863
         512    3.148    8.979    6.324      0.381
        1024    3.079    9.659    6.655      0.461
        2048    3.809   11.288    7.645      0.428
        4096    5.181   14.052    9.559      0.565
        8192    8.281   18.844   13.830      0.975
       16384   15.848   27.646   21.691      1.337
       32768   21.922   39.456   31.782      1.743
       65536   36.499   69.583   55.524      3.959
      131072   88.287   97.279   92.802      1.298
      262144  152.497  166.167  158.760      9.482
      524288  332.342  347.742  338.839      2.425
     1048576  453.261  486.567  470.339      8.603
     2097152  778.914  864.101  827.881      8.901
     4194304 1368.881 1500.696 1431.112     13.682

TEST bcast:
    :root=1
     msgsize      min      max      avg      sigma  (in microseconds)
           0    0.024    0.027    0.024      0.005
           1    0.728    7.606    4.217      0.369
           2    0.618    7.568    4.138      0.325
           4    0.637    7.529    4.124      0.287
           8    0.675    7.453    4.117      0.276
          16    0.651    7.460    4.114      0.309
          32    0.721    7.570    4.148      0.301
          64    0.632    7.573    4.154      0.297
         128    0.559    7.513    4.185      0.300
         256    3.231    8.348    5.392      0.290
         512    3.536    8.858    5.733      0.297
        1024    3.632    9.735    6.249      0.343
        2048    4.056   11.072    7.055      0.500
        4096    4.809   13.727    8.704      0.645
        8192    8.584   19.349   13.434      0.756
       16384   13.921   25.918   21.324      0.959
       32768   21.643   38.869   31.534      1.374
       65536   34.536   64.359   53.487      3.525
      131072   88.825   96.685   92.535      1.346
      262144  147.846  168.210  156.677     10.115
      524288  365.246  408.866  389.087     244.849
     1048576  476.797  496.886  490.213     129.581
     2097152  679.991  702.364  688.366     60.602
     4194304 1123.676 1199.528 1160.844     358.724

```
#### This PR
```
TEST bcast:
    :root=0
     msgsize      min      max      avg      sigma  (in microseconds)
           0    0.024    0.025    0.024      0.003
           1    0.284    6.246    4.463      0.384
           2    0.274    6.115    4.432      0.277
           4    0.274    6.400    4.527      0.304
           8    0.273    6.245    4.444      0.295
          16    0.272    6.174    4.462      0.284
          32    0.273    6.314    4.485      0.286
          64    0.273    6.352    4.563      0.288
         128    0.273    6.465    4.583      0.298
         256    5.004   14.651    9.816      0.271
         512    5.003   15.112   10.111      0.360
        1024    5.194   16.181   10.714      0.426
        2048    5.383   17.896   11.818      0.461
        4096    5.675   20.559   13.372      0.547
        8192    9.117   31.716   21.610      0.673
       16384   14.081   52.932   35.729      0.842
       32768   25.681   51.923   40.205      0.909
       65536   32.539   79.681   60.213      0.976
      131072   88.475   97.878   93.484      1.357
      262144  152.020  163.570  158.207      9.927
      524288  332.817  345.644  338.733      2.078
     1048576  431.397  443.785  437.692      6.365
     2097152  760.351  782.076  772.526     53.734
     4194304 1345.219 1386.821 1366.342     22.916

TEST bcast:
    :root=1
     msgsize      min      max      avg      sigma  (in microseconds)
           0    0.024    0.031    0.024      0.003
           1    0.446    7.305    4.105      0.284
           2    0.578    7.592    4.173      0.325
           4    0.569    7.534    4.137      0.297
           8    0.577    7.494    4.136      0.317
          16    0.562    7.444    4.112      0.268
          32    0.574    7.469    4.145      0.280
          64    0.542    7.447    4.129      0.269
         128    0.547    7.484    4.164      0.320
         256    3.674   15.096    9.457      0.310
         512    3.795   15.453    9.716      0.330
        1024    4.087   16.468   10.339      0.379
        2048    4.605   18.407   11.553      0.534
        4096    5.567   20.858   13.136      0.649
        8192    9.326   32.901   22.096      0.862
       16384    9.643   53.886   34.692      1.075
       32768   22.440   51.148   38.893      1.022
       65536   31.826   80.056   59.780      0.941
      131072   89.712   97.311   92.808      1.355
      262144  151.675  166.279  157.243     10.696
      524288  366.917  411.135  397.077     284.970
     1048576  510.357  528.699  516.871     323.150
     2097152  695.428  709.956  702.151     115.961
     4194304 1148.820 1202.485 1181.695     345.061
```

The bcast latency is measured after a barrier, and due to different processes exiting barrier unevenly there are a difference between `root=0` and `root=1` that is not due to the bcast algorithm. However, I think the data is sufficient to show the new algorithm is *not slower* than the `main` branch.